src/transforms/projectionpursuit.jl

```diff
@@ -0,0 +1,204 @@
+distributions, or if the number of iterations surpasses `maxiter` (the maximum 
+number of iterations allowed).
+
+
+
+# Examples
+
+```julia
+ProjectionPursuit()
+ProjectionPursuit(deg=10)
+ProjectionPursuit(perc=.85, n=50)
+ProjectionPursuit(tol=1e-4, maxiter=250, deg=5, perc=.95, n=100)
+```
+
+"""
+
+struct ProjectionPursuit{T} <: Transform
+  tol::T
+  maxiter::Int
+  deg::Int
+  perc::T
+  n::Int
+end
+
+ProjectionPursuit(;tol=1e-6, maxiter=100, deg=5, perc=.9, n=100) =
+  ProjectionPursuit{typeof(tol)}(tol, maxiter, deg, perc, n)
+
+isrevertible(::Type{<:ProjectionPursuit}) = true
+
+#Transforms a row of random variables into a convex combination 
+#random variable with values in [-1,1] and standard normal distribution.
+function rscore(Z, α)
+  ᾱ = (1/norm(α)) .* α .* (norm(α)!=1) + α .* (norm(α)==1)
+  X = Z * ᾱ
+  2 * cdf(Normal(), X) .- 1
+end
+
+function pindex(transform, Z, α)
+  X = rscore(Z, α)
+  I = (3/2) * mean(X)^2
+  if transform.deg > 1
+    Pⱼ₋₂, Pⱼ₋₁ = ones(size(X)), X
+    for j = 2:transform.deg
+      Pⱼ₋₂, Pⱼ₋₁ = 
+        Pⱼ₋₁, (1/j) * ((2j-1) * X .* Pⱼ₋₁ - (j-1) * Pⱼ₋₂)
+      I += ((2j+1)/2) * (mean(Pⱼ₋₁))^2
+    end
+  end
+  return I
+end
+
+function optimaxis(transform, Z)
+  _, q = size(Z)
+  Iₘₐₓ, iₘₐₓ = 0, 0
+  for i = 1:q
+    eᵢ = [1. * (j==i) for j = 1:q]
+    I = pindex(transform, Z, eᵢ)
+    if I > Iₘₐₓ
+      Iₘₐₓ, iₘₐₓ = I, i
+    end
+  end
+  return [1. * (iₘₐₓ==i) for i = 1:q]
+end
+
+function optimdiag(transform, Z, ᾱ)
+  l = length(ᾱ)
+  I₀ = pindex(transform, Z, ᾱ)
+  for i in 1:l
+    eᵢ = [1. * (j==i) for j = 1:l]
+    f₊ = pindex(transform, Z, (1/√(2+2ᾱ[i])) * (ᾱ + eᵢ))
+    f₋ = pindex(transform, Z, (1/√(2-2ᾱ[i])) * (ᾱ - eᵢ)) * (ᾱ[i] != 1)
+    f = (f₊ > f₋) * f₊ + (f₊ ≤ f₋) * f₋ 
+    s = (f₊ > f₋) * 1 + (f₊ ≤ f₋) * (-1)
+    if f > pindex(transform, Z, ᾱ)
+      ᾱ = (1/√(2+2s*ᾱ[i])) * (ᾱ + s*eᵢ)
+    end
+  end
+  if pindex(transform, Z, ᾱ) ≥ I₀
+    return ᾱ
+  else
+    return optimaxis(transform, Z)
+  end
+end
+
+function optimgen(transform, Z, α₀)
+  f(α) = -pindex(transform, Z, α)
+  op = optimize(f, α₀)
+  Optim.minimizer(op)
+end
+
+function optim(transform, Z)
+  α_axis = optimaxis(transform, Z)
+  α_diag = optimdiag(transform, Z, α_axis)
+  optimgen(transform, Z, α_diag)  
+end
+
+#Transforms the cdf of a convex combination of r.v. to a standard normal cdf.
+function structure_removal(transform, Z, α)
+  l = length(α)
+  U = qr(hcat(α, rand(l, l-1)))
+  if norm(diag(U.R)) < transform.tol
+    return print("Error: the random matrix is singular. Try again")
+  end
+  T = Z * U.Q
+  table = Tables.table(T)
+  newtable, _ = apply(Quantile(), table)
+  T[:,1] = newtable[:,1]
+  return T * U.Q'
+end
+
+#Gives the array that will be used to stop the loop of structure 
+#removal by computing the projection index on k samples of multinomial 
+#distribution (size of sample kxn) along q random directions.
+function gauss_pmatrix(transform, N, q)
```

Comment:
this function is returning I and Q, but these variables aren't used anywhere in the main iteration loop, can you please double check?

Reply:
They are used to define the stopcriteria

Reply:
I just reviewed the code and the Q is necessary: The columns of Q define the directions at which the gaussian samples pindex was calculated. So when comparing with the matrix I, the table introduced uses Q to know the directions to be compared.

Reply:
What about always projecting the standard indices along the major axes? Would that save us from having to pass Q around and always use the identity for example? We could reduce the number of allocations because in Julia I is a lazy object.

Reply:
That isn't possible because the first column is the alpha, and the matrix must be orthogonal.

Reply:
That's a good idea. I'll change it

Reply:
I think you need to assess what is the most efficient option. You can either fix the major axes for the stopping criteria and compute the scores for the dataset along these axes as well besides the rotated basis, or you can use the same basis or rotation you currently have to compare with the Gaussian case. The point is that you have some freedom to choose the basis of comparison with the Gaussian case.